File: tests/translation_status_command_integration_test.rs
use chrono::Duration;
use lib::subtitle::model::{Subtitle, SubtitleIndex, SubtitleText, SubtitleTimestamp};
use lib::subtitle::ports::SubtitleService;
use lib::subtitle::service::SubRipService;
use lib::translation_status::service;
use std::fs;
use tempfile::TempDir;

fn make_test_subtitle(index: u32, start_ms: i64, end_ms: i64, text: &str) -> Subtitle {
    Subtitle::new(
        SubtitleIndex::try_new(index).unwrap(),
        SubtitleTimestamp::try_new(Duration::milliseconds(start_ms)).unwrap(),
        SubtitleTimestamp::try_new(Duration::milliseconds(end_ms)).unwrap(),
        SubtitleText::try_new(text.to_string()).unwrap(),
    )
    .unwrap()
}

fn create_srt_file(dir: &TempDir, filename: &str, subtitles: Vec<Subtitle>) -> String {
    let file_path = dir.path().join(filename);
    let service = SubRipService::new(dir.path().to_path_buf());

    // Create empty file first
    fs::write(&file_path, "").unwrap();

    // Add each subtitle
    for sub in subtitles {
        service
            .add(filename, sub.start_time, sub.end_time, sub.text)
            .unwrap();
    }

    file_path.to_str().unwrap().to_string()
}

#[test]
fn test_empty_translation_file() {
    let temp_dir = TempDir::new().unwrap();

    let ref_subs = vec![
        make_test_subtitle(1, 1000, 2000, "English 1"),
        make_test_subtitle(2, 2000, 3000, "English 2"),
        make_test_subtitle(3, 3000, 4000, "English 3"),
    ];

    let translation_subs = vec![];

    create_srt_file(&temp_dir, "reference.srt", ref_subs.clone());
    create_srt_file(&temp_dir, "translation.srt", translation_subs.clone());

    let report = service::check_translation_status(
        ref_subs,
        "reference.srt".to_string(),
        translation_subs,
        "translation.srt".to_string(),
        50,
    );

    assert_eq!(report.total_count, 3);
    assert_eq!(report.translated_count, 0);
    assert_eq!(report.missing_count, 3);
    assert_eq!(report.progress_percentage(), 0.0);
    assert!(!report.is_complete());

    assert!(report.next_chunk.is_some());
    let chunk = report.next_chunk.unwrap();
    assert_eq!(chunk.start_index, 1);
    assert_eq!(chunk.end_index, 3);
}

#[test]
fn test_partial_translation() {
    let temp_dir = TempDir::new().unwrap();

    let ref_subs = vec![
        make_test_subtitle(1, 1000, 2000, "English 1"),
        make_test_subtitle(2, 2000, 3000, "English 2"),
        make_test_subtitle(3, 3000, 4000, "English 3"),
        make_test_subtitle(4, 4000, 5000, "English 4"),
        make_test_subtitle(5, 5000, 6000, "English 5"),
    ];

    let translation_subs = vec![
        make_test_subtitle(1, 1000, 2000, "Russian 1"),
        make_test_subtitle(2, 2000, 3000, "Russian 2"),
    ];

    create_srt_file(&temp_dir, "reference.srt", ref_subs.clone());
    create_srt_file(&temp_dir, "translation.srt", translation_subs.clone());

    let report = service::check_translation_status(
        ref_subs,
        "reference.srt".to_string(),
        translation_subs,
        "translation.srt".to_string(),
        50,
    );

    assert_eq!(report.total_count, 5);
    assert_eq!(report.translated_count, 2);
    assert_eq!(report.missing_count, 3);
    assert_eq!(report.progress_percentage(), 40.0);
    assert!(!report.is_complete());

    assert!(report.next_chunk.is_some());
    let chunk = report.next_chunk.unwrap();
    assert_eq!(chunk.start_index, 3);
    assert_eq!(chunk.end_index, 5);
}

#[test]
fn test_complete_translation() {
    let temp_dir = TempDir::new().unwrap();

    let ref_subs = vec![
        make_test_subtitle(1, 1000, 2000, "English 1"),
        make_test_subtitle(2, 2000, 3000, "English 2"),
        make_test_subtitle(3, 3000, 4000, "English 3"),
    ];

    let translation_subs = vec![
        make_test_subtitle(1, 1000, 2000, "Russian 1"),
        make_test_subtitle(2, 2000, 3000, "Russian 2"),
        make_test_subtitle(3, 3000, 4000, "Russian 3"),
    ];

    create_srt_file(&temp_dir, "reference.srt", ref_subs.clone());
    create_srt_file(&temp_dir, "translation.srt", translation_subs.clone());

    let report = service::check_translation_status(
        ref_subs,
        "reference.srt".to_string(),
        translation_subs,
        "translation.srt".to_string(),
        50,
    );

    assert_eq!(report.total_count, 3);
    assert_eq!(report.translated_count, 3);
    assert_eq!(report.missing_count, 0);
    assert_eq!(report.progress_percentage(), 100.0);
    assert!(report.is_complete());
    assert!(report.next_chunk.is_none());
}

#[test]
fn test_translation_with_gaps() {
    let temp_dir = TempDir::new().unwrap();

    let ref_subs = vec![
        make_test_subtitle(1, 1000, 2000, "English 1"),
        make_test_subtitle(2, 2000, 3000, "English 2"),
        make_test_subtitle(5, 5000, 6000, "English 5"),
        make_test_subtitle(10, 10000, 11000, "English 10"),
    ];

    let translation_subs = vec![
        make_test_subtitle(1, 1000, 2000, "Russian 1"),
        make_test_subtitle(10, 10000, 11000, "Russian 10"),
    ];

    create_srt_file(&temp_dir, "reference.srt", ref_subs.clone());
    create_srt_file(&temp_dir, "translation.srt", translation_subs.clone());

    let report = service::check_translation_status(
        ref_subs,
        "reference.srt".to_string(),
        translation_subs,
        "translation.srt".to_string(),
        50,
    );

    assert_eq!(report.total_count, 4);
    assert_eq!(report.translated_count, 2);
    assert_eq!(report.missing_count, 2);
    assert_eq!(report.progress_percentage(), 50.0);

    assert!(report.next_chunk.is_some());
    let chunk = report.next_chunk.unwrap();
    assert_eq!(chunk.start_index, 2);
    assert_eq!(chunk.end_index, 5);
}

#[test]
fn test_custom_chunk_size() {
    let temp_dir = TempDir::new().unwrap();

    let ref_subs: Vec<Subtitle> = (1..=100)
        .map(|i| {
            make_test_subtitle(
                i,
                i as i64 * 1000,
                (i as i64 + 1) * 1000,
                &format!("Text {}", i),
            )
        })
        .collect();

    let translation_subs = vec![];

    create_srt_file(&temp_dir, "reference.srt", ref_subs.clone());
    create_srt_file(&temp_dir, "translation.srt", translation_subs.clone());

    // Test with chunk size of 25
    let report = service::check_translation_status(
        ref_subs.clone(),
        "reference.srt".to_string(),
        translation_subs.clone(),
        "translation.srt".to_string(),
        25,
    );

    assert!(report.next_chunk.is_some());
    let chunk = report.next_chunk.unwrap();
    assert_eq!(chunk.start_index, 1);
    assert_eq!(chunk.end_index, 25);

    // Test with chunk size of 10
    let report = service::check_translation_status(
        ref_subs,
        "reference.srt".to_string(),
        translation_subs,
        "translation.srt".to_string(),
        10,
    );

    assert!(report.next_chunk.is_some());
    let chunk = report.next_chunk.unwrap();
    assert_eq!(chunk.start_index, 1);
    assert_eq!(chunk.end_index, 10);
}

#[test]
fn test_chunk_size_larger_than_remaining() {
    let temp_dir = TempDir::new().unwrap();

    let ref_subs = vec![
        make_test_subtitle(1, 1000, 2000, "English 1"),
        make_test_subtitle(2, 2000, 3000, "English 2"),
        make_test_subtitle(3, 3000, 4000, "English 3"),
    ];

    let translation_subs = vec![make_test_subtitle(1, 1000, 2000, "Russian 1")];

    create_srt_file(&temp_dir, "reference.srt", ref_subs.clone());
    create_srt_file(&temp_dir, "translation.srt", translation_subs.clone());

    let report = service::check_translation_status(
        ref_subs,
        "reference.srt".to_string(),
        translation_subs,
        "translation.srt".to_string(),
        100,
    );

    assert!(report.next_chunk.is_some());
    let chunk = report.next_chunk.unwrap();
    assert_eq!(chunk.start_index, 2);
    assert_eq!(chunk.end_index, 3); // Capped at last missing index
}

File: internal/cli/cmd/translation_status_command_integration_test.go
package cmd

import (
	"os"
	"path/filepath"
	"testing"
	"time"

	"github.com/lebe-dev/submarine/pkg/subtitle"
	"github.com/lebe-dev/submarine/pkg/translationstatus"
)

// makeTestSubtitle mirrors the Rust `make_test_subtitle` helper.
func makeTestSubtitle(t *testing.T, index uint32, startMs, endMs int64, text string) subtitle.Subtitle {
	t.Helper()
	idx, err := subtitle.NewSubtitleIndex(index)
	if err != nil {
		t.Fatalf("invalid index: %v", err)
	}
	start, err := subtitle.NewSubtitleTimestamp(time.Duration(startMs) * time.Millisecond)
	if err != nil {
		t.Fatalf("invalid start: %v", err)
	}
	end, err := subtitle.NewSubtitleTimestamp(time.Duration(endMs) * time.Millisecond)
	if err != nil {
		t.Fatalf("invalid end: %v", err)
	}
	txt, err := subtitle.NewSubtitleText(text)
	if err != nil {
		t.Fatalf("invalid text: %v", err)
	}
	sub, err := subtitle.NewSubtitle(idx, start, end, txt)
	if err != nil {
		t.Fatalf("invalid subtitle: %v", err)
	}
	return sub
}

// createSrtFile mirrors the Rust `create_srt_file` helper: writes an empty SRT
// file in dir, then adds each subtitle through the SubRipService. Returns the
// full path to the created file.
func createSrtFile(t *testing.T, dir, filename string, subtitles []subtitle.Subtitle) string {
	t.Helper()
	filePath := filepath.Join(dir, filename)
	service := subtitle.NewSubRipService(dir)

	if err := os.WriteFile(filePath, []byte(""), 0o644); err != nil {
		t.Fatalf("failed to write %s: %v", filePath, err)
	}

	for _, sub := range subtitles {
		if _, err := service.Add(filename, sub.StartTime, sub.EndTime, sub.Text); err != nil {
			t.Fatalf("failed to add subtitle: %v", err)
		}
	}

	return filePath
}

// TestIntegrationEmptyTranslationFile ports `test_empty_translation_file`.
func TestIntegrationEmptyTranslationFile(t *testing.T) {
	dir := t.TempDir()

	refSubs := []subtitle.Subtitle{
		makeTestSubtitle(t, 1, 1000, 2000, "English 1"),
		makeTestSubtitle(t, 2, 2000, 3000, "English 2"),
		makeTestSubtitle(t, 3, 3000, 4000, "English 3"),
	}
	var translationSubs []subtitle.Subtitle

	createSrtFile(t, dir, "reference.srt", refSubs)
	createSrtFile(t, dir, "translation.srt", translationSubs)

	report := translationstatus.CheckTranslationStatus(
		refSubs, "reference.srt", translationSubs, "translation.srt", 50,
	)

	if report.TotalCount != 3 {
		t.Errorf("TotalCount = %d, want 3", report.TotalCount)
	}
	if report.TranslatedCount != 0 {
		t.Errorf("TranslatedCount = %d, want 0", report.TranslatedCount)
	}
	if report.MissingCount != 3 {
		t.Errorf("MissingCount = %d, want 3", report.MissingCount)
	}
	if report.ProgressPercentage() != 0.0 {
		t.Errorf("ProgressPercentage = %v, want 0.0", report.ProgressPercentage())
	}
	if report.IsComplete() {
		t.Error("IsComplete = true, want false")
	}
	if report.NextChunk == nil {
		t.Fatal("NextChunk should be set")
	}
	if report.NextChunk.StartIndex != 1 {
		t.Errorf("chunk.StartIndex = %d, want 1", report.NextChunk.StartIndex)
	}
	if report.NextChunk.EndIndex != 3 {
		t.Errorf("chunk.EndIndex = %d, want 3", report.NextChunk.EndIndex)
	}
}

// TestIntegrationPartialTranslation ports `test_partial_translation`.
func TestIntegrationPartialTranslation(t *testing.T) {
	dir := t.TempDir()

	refSubs := []subtitle.Subtitle{
		makeTestSubtitle(t, 1, 1000, 2000, "English 1"),
		makeTestSubtitle(t, 2, 2000, 3000, "English 2"),
		makeTestSubtitle(t, 3, 3000, 4000, "English 3"),
		makeTestSubtitle(t, 4, 4000, 5000, "English 4"),
		makeTestSubtitle(t, 5, 5000, 6000, "English 5"),
	}
	translationSubs := []subtitle.Subtitle{
		makeTestSubtitle(t, 1, 1000, 2000, "Russian 1"),
		makeTestSubtitle(t, 2, 2000, 3000, "Russian 2"),
	}

	createSrtFile(t, dir, "reference.srt", refSubs)
	createSrtFile(t, dir, "translation.srt", translationSubs)

	report := translationstatus.CheckTranslationStatus(
		refSubs, "reference.srt", translationSubs, "translation.srt", 50,
	)

	if report.TotalCount != 5 {
		t.Errorf("TotalCount = %d, want 5", report.TotalCount)
	}
	if report.TranslatedCount != 2 {
		t.Errorf("TranslatedCount = %d, want 2", report.TranslatedCount)
	}
	if report.MissingCount != 3 {
		t.Errorf("MissingCount = %d, want 3", report.MissingCount)
	}
	if report.ProgressPercentage() != 40.0 {
		t.Errorf("ProgressPercentage = %v, want 40.0", report.ProgressPercentage())
	}
	if report.IsComplete() {
		t.Error("IsComplete = true, want false")
	}
	if report.NextChunk == nil {
		t.Fatal("NextChunk should be set")
	}
	if report.NextChunk.StartIndex != 3 {
		t.Errorf("chunk.StartIndex = %d, want 3", report.NextChunk.StartIndex)
	}
	if report.NextChunk.EndIndex != 5 {
		t.Errorf("chunk.EndIndex = %d, want 5", report.NextChunk.EndIndex)
	}
}

// TestIntegrationCompleteTranslation ports `test_complete_translation`.
func TestIntegrationCompleteTranslation(t *testing.T) {
	dir := t.TempDir()

	refSubs := []subtitle.Subtitle{
		makeTestSubtitle(t, 1, 1000, 2000, "English 1"),
		makeTestSubtitle(t, 2, 2000, 3000, "English 2"),
		makeTestSubtitle(t, 3, 3000, 4000, "English 3"),
	}
	translationSubs := []subtitle.Subtitle{
		makeTestSubtitle(t, 1, 1000, 2000, "Russian 1"),
		makeTestSubtitle(t, 2, 2000, 3000, "Russian 2"),
		makeTestSubtitle(t, 3, 3000, 4000, "Russian 3"),
	}

	createSrtFile(t, dir, "reference.srt", refSubs)
	createSrtFile(t, dir, "translation.srt", translationSubs)

	report := translationstatus.CheckTranslationStatus(
		refSubs, "reference.srt", translationSubs, "translation.srt", 50,
	)

	if report.TotalCount != 3 {
		t.Errorf("TotalCount = %d, want 3", report.TotalCount)
	}
	if report.TranslatedCount != 3 {
		t.Errorf("TranslatedCount = %d, want 3", report.TranslatedCount)
	}
	if report.MissingCount != 0 {
		t.Errorf("MissingCount = %d, want 0", report.MissingCount)
	}
	if report.ProgressPercentage() != 100.0 {
		t.Errorf("ProgressPercentage = %v, want 100.0", report.ProgressPercentage())
	}
	if !report.IsComplete() {
		t.Error("IsComplete = false, want true")
	}
	if report.NextChunk != nil {
		t.Errorf("NextChunk should be nil, got %+v", report.NextChunk)
	}
}

// TestIntegrationTranslationWithGaps ports `test_translation_with_gaps`.
func TestIntegrationTranslationWithGaps(t *testing.T) {
	dir := t.TempDir()

	refSubs := []subtitle.Subtitle{
		makeTestSubtitle(t, 1, 1000, 2000, "English 1"),
		makeTestSubtitle(t, 2, 2000, 3000, "English 2"),
		makeTestSubtitle(t, 5, 5000, 6000, "English 5"),
		makeTestSubtitle(t, 10, 10000, 11000, "English 10"),
	}
	translationSubs := []subtitle.Subtitle{
		makeTestSubtitle(t, 1, 1000, 2000, "Russian 1"),
		makeTestSubtitle(t, 10, 10000, 11000, "Russian 10"),
	}

	createSrtFile(t, dir, "reference.srt", refSubs)
	createSrtFile(t, dir, "translation.srt", translationSubs)

	report := translationstatus.CheckTranslationStatus(
		refSubs, "reference.srt", translationSubs, "translation.srt", 50,
	)

	if report.TotalCount != 4 {
		t.Errorf("TotalCount = %d, want 4", report.TotalCount)
	}
	if report.TranslatedCount != 2 {
		t.Errorf("TranslatedCount = %d, want 2", report.TranslatedCount)
	}
	if report.MissingCount != 2 {
		t.Errorf("MissingCount = %d, want 2", report.MissingCount)
	}
	if report.ProgressPercentage() != 50.0 {
		t.Errorf("ProgressPercentage = %v, want 50.0", report.ProgressPercentage())
	}
	if report.NextChunk == nil {
		t.Fatal("NextChunk should be set")
	}
	if report.NextChunk.StartIndex != 2 {
		t.Errorf("chunk.StartIndex = %d, want 2", report.NextChunk.StartIndex)
	}
	if report.NextChunk.EndIndex != 2 {
		t.Errorf("chunk.EndIndex = %d, want 2", report.NextChunk.EndIndex)
	}
}

// TestIntegrationCustomChunkSize ports `test_custom_chunk_size`.
func TestIntegrationCustomChunkSize(t *testing.T) {
	dir := t.TempDir()

	refSubs := make([]subtitle.Subtitle, 0, 100)
	for i := int64(1); i <= 100; i++ {
		refSubs = append(refSubs, makeTestSubtitle(t, uint32(i), i*1000, (i+1)*1000, "Text"))
	}
	var translationSubs []subtitle.Subtitle

	createSrtFile(t, dir, "reference.srt", refSubs)
	createSrtFile(t, dir, "translation.srt", translationSubs)

	// chunk size 25
	report := translationstatus.CheckTranslationStatus(
		refSubs, "reference.srt", translationSubs, "translation.srt", 25,
	)
	if report.NextChunk == nil {
		t.Fatal("NextChunk should be set (chunk size 25)")
	}
	if report.NextChunk.StartIndex != 1 {
		t.Errorf("chunk.StartIndex = %d, want 1", report.NextChunk.StartIndex)
	}
	if report.NextChunk.EndIndex != 25 {
		t.Errorf("chunk.EndIndex = %d, want 25", report.NextChunk.EndIndex)
	}

	// chunk size 10
	report = translationstatus.CheckTranslationStatus(
		refSubs, "reference.srt", translationSubs, "translation.srt", 10,
	)
	if report.NextChunk == nil {
		t.Fatal("NextChunk should be set (chunk size 10)")
	}
	if report.NextChunk.StartIndex != 1 {
		t.Errorf("chunk.StartIndex = %d, want 1", report.NextChunk.StartIndex)
	}
	if report.NextChunk.EndIndex != 10 {
		t.Errorf("chunk.EndIndex = %d, want 10", report.NextChunk.EndIndex)
	}
}

// TestIntegrationChunkSizeLargerThanRemaining ports
// `test_chunk_size_larger_than_remaining`.
func TestIntegrationChunkSizeLargerThanRemaining(t *testing.T) {
	dir := t.TempDir()

	refSubs := []subtitle.Subtitle{
		makeTestSubtitle(t, 1, 1000, 2000, "English 1"),
		makeTestSubtitle(t, 2, 2000, 3000, "English 2"),
		makeTestSubtitle(t, 3, 3000, 4000, "English 3"),
	}
	translationSubs := []subtitle.Subtitle{
		makeTestSubtitle(t, 1, 1000, 2000, "Russian 1"),
	}

	createSrtFile(t, dir, "reference.srt", refSubs)
	createSrtFile(t, dir, "translation.srt", translationSubs)

	report := translationstatus.CheckTranslationStatus(
		refSubs, "reference.srt", translationSubs, "translation.srt", 100,
	)

	if report.NextChunk == nil {
		t.Fatal("NextChunk should be set")
	}
	if report.NextChunk.StartIndex != 2 {
		t.Errorf("chunk.StartIndex = %d, want 2", report.NextChunk.StartIndex)
	}
	if report.NextChunk.EndIndex != 3 {
		t.Errorf("chunk.EndIndex = %d, want 3 (capped at last missing index)", report.NextChunk.EndIndex)
	}
}
